trillian/ctfe/services.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ctfe
+
+import (
+	"context"
+	"crypto/sha256"
+
+	"github.com/google/certificate-transparency-go/trillian/ctfe/cache"
+	"github.com/google/certificate-transparency-go/trillian/ctfe/storage"
+	"k8s.io/klog/v2"
+)
+
+type IssuanceChainStorageBackend string
+
+const (
+	IssuanceChainStorageBackendTrillianGRPC = IssuanceChainStorageBackend("TrillianGRPC")
+	IssuanceChainStorageBackendCTFE         = IssuanceChainStorageBackend("CTFE")
+)
+
+type IssuanceChainService struct {
+	storageBackend IssuanceChainStorageBackend
+	storage        storage.IssuanceChainStorage
+	cache          cache.IssuanceChainCache
+}
+
+func NewIssuanceChainService(ctx context.Context, storageBackend IssuanceChainStorageBackend, s storage.IssuanceChainStorage, c cache.IssuanceChainCache) *IssuanceChainService {
+	service := &IssuanceChainService{
+		storageBackend: storageBackend,
+		storage:        s,
+		cache:          c,
+	}
+
+	return service
+}
+
+func (s *IssuanceChainService) IsTrillianGRPCStorageBackendEnabled() bool {
+	return s.storageBackend == IssuanceChainStorageBackendTrillianGRPC
+}
+
+func (s *IssuanceChainService) IsCTFEStorageBackendEnabled() bool {
+	return s.storageBackend == IssuanceChainStorageBackendCTFE
+}
```

Comment:
Do these methods need to be here? It would be a cleaner design if this service didn't need to know about the `storageBackend`, as it doesn't look like it's needed for its implementation.

Reply:
Let me remove these methods and add them back when needed in other pull requests.

I would like to put all issuance chain related business logic or methods in this service layer. For example, what's the storage backend of issuance chain? Ask the issuance chain service.

Reply:
Let's discuss when these need to be re-added. The goal of interfaces is to hide that there are different implementations from client code. If this is successfully done then no other code should care about which implementation is being used. If client code needs to care, then this is a sign that there's a leaky abstraction.

But let's discuss when the client code that needs these methods is added.

Reply:
Sure, thanks!